src/librustc_mir/transform/const_prop.rs

```diff
@@ -988,8 +1023,103 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
+                //                         trace!("ScalarMaybeUndef of the OpTy: {:?}", &scalar_mu);
+                //                         if let ScalarMaybeUndef::Scalar(scalar) = scalar_mu {
+                //                             trace!("Scalar of the ScalarMU: {:?}", &scalar);
+                //                             let place_type = op_ty.layout.ty;
+                //                             trace!("Type of the OpTy: {:?}", &place_type);
+                //                             *opr = self.operand_from_scalar(
+                //                                 *scalar,
+                //                                 place_type,
+                //                                 source_info.span,
+                //                             );
+                //                         }
+                //                     }
+                //                 }
+                //             }
+                //         }
+                //     }
+                // }
+                // Second attempt, using wesleywiser's suggested functions
+                // for opr in args {
+                //     if let Some(op_ty) = self.eval_operand(&opr, source_info) {
+                //         trace!("OpTy of the place: {:?}", &op_ty);
+                //         // Always safe to unwrap: look at `read_scalar` in `interpret/operand.rs`
+                //         let scalar_mu = self.ecx.read_scalar(op_ty).unwrap();
+                //         trace!("ScalarMaybeUndef of the OpTy: {:?}", &scalar_mu);
+                //         if let ScalarMaybeUndef::Scalar(scalar) = scalar_mu {
+                //             trace!("Scalar of the ScalarMU: {:?}", &scalar);
+                //             let operand_type = opr.ty(&self.local_decls, self.tcx);
+                //             trace!("Type of the OpTy: {:?}", &operand_type);
+                //             *opr = self.operand_from_scalar(scalar, operand_type, source_info.span);
+                //         }
+                //     }
+                // }
+                // Third attempt, while wesleywiser comes back
+                for opr in args {
+                    if let Some(op_ty) = self.eval_operand(&opr, source_info) {
+                        trace!("OpTy of the place: {:?}", &op_ty);
+                        if self.should_const_prop(op_ty) {
+                            if let interpret::Operand::Immediate(immediate) = *op_ty {
+                                // Only Immediate from here on
+                                // FIXME(felix91gr): This only works for Immediate. Investigate
+                                // if it could be made to work for Indirect as well.
+                                trace!("Immediate of the OpTy: {:?}", &immediate);
+                                if let interpret::Immediate::Scalar(scalar_mu) = immediate {
```

Comment:
You can nest these patterns:

```rust
if let interpret::Immediate::Scalar(ScalarMaybeUndef::Scalar(scalar)) = immediate {
```

Reply:
Ohhh, I love it!

Reply:
I wanted to use this, but I want to leave the super polish of these branches to the next PR, since we'll be changing this piece of code anyways.